File: benchmark/languages/c/work/day09/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef long long ll;

static int cmp_int(const void *a, const void *b) {
    int x = *(int*)a, y = *(int*)b;
    return x < y ? -1 : x > y ? 1 : 0;
}

static int unique_arr(int *arr, int n) {
    if (n == 0) return 0;
    int k = 0;
    for (int i = 0; i < n; i++) {
        if (i == 0 || arr[i] != arr[i-1]) arr[k++] = arr[i];
    }
    return k;
}

static int find_idx(int *arr, int n, int val) {
    int lo = 0, hi = n - 1;
    while (lo <= hi) {
        int mid = (lo + hi) / 2;
        if (arr[mid] == val) return mid;
        else if (arr[mid] < val) lo = mid + 1;
        else hi = mid - 1;
    }
    return -1;
}

// Check if point (px, py) is inside or on boundary of rectilinear polygon.
// Polygon edges connect consecutive vertices (either horiz or vert).
// Uses ray casting to the LEFT; vertical edges use half-open interval [ymin, ymax).
static int is_valid(int px, int py, int *xs, int *ys, int n) {
    // Check boundary first
    for (int i = 0; i < n; i++) {
        int x1 = xs[i], y1 = ys[i];
        int x2 = xs[(i+1)%n], y2 = ys[(i+1)%n];
        if (y1 == y2) {
            // Horizontal edge
            int xmin = x1 < x2 ? x1 : x2;
            int xmax = x1 > x2 ? x1 : x2;
            if (py == y1 && px >= xmin && px <= xmax) return 1;
        } else {
            // Vertical edge at x = x1 (= x2)
            int ymin = y1 < y2 ? y1 : y2;
            int ymax = y1 > y2 ? y1 : y2;
            if (px == x1 && py >= ymin && py <= ymax) return 1;
        }
    }
    // Ray to the left: count vertical edges with x1 <= px crossing y=py
    // Use half-open interval [ymin, ymax) to avoid double-counting at vertices
    int crossings = 0;
    for (int i = 0; i < n; i++) {
        int x1 = xs[i], y1 = ys[i];
        int x2 = xs[(i+1)%n], y2 = ys[(i+1)%n];
        if (y1 != y2) {
            // Vertical edge at x = x1
            int ymin = y1 < y2 ? y1 : y2;
            int ymax = y1 > y2 ? y1 : y2;
            if (py >= ymin && py < ymax && x1 <= px) {
                crossings++;
            }
        }
        (void)x2; // suppress unused warning (x2 == x1 for vertical edge)
    }
    return crossings % 2 == 1;
}

int main(int argc, char *argv[]) {
    int part = 0;
    if (argc > 1) part = atoi(argv[1]);

    FILE *f = fopen("input.txt", "r");
    if (!f) { perror("input.txt"); return 1; }

    int *xs = malloc(10000 * sizeof(int));
    int *ys = malloc(10000 * sizeof(int));
    int n = 0;
    int x, y;
    while (fscanf(f, "%d,%d", &x, &y) == 2) {
        xs[n] = x; ys[n] = y; n++;
    }
    fclose(f);

    // ---- Part 1 ----
    // Find largest rectangle with two red tiles as opposite corners.
    // Area = (|dx|+1) * (|dy|+1). Check all O(N^2) pairs.
    ll best1 = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            ll dx = xs[j] - xs[i]; if (dx < 0) dx = -dx;
            ll dy = ys[j] - ys[i]; if (dy < 0) dy = -dy;
            ll area = (dx + 1) * (dy + 1);
            if (area > best1) best1 = area;
        }
    }

    if (part == 1) {
        printf("Part 1: %lld\n", best1);
        free(xs); free(ys);
        return 0;
    }

    // ---- Part 2 ----
    // The red tiles form a rectilinear polygon (consecutive tiles share row/col).
    // Valid tiles = boundary + interior. Rectangle must have red corners and all
    // tiles inside must be valid.
    //
    // Use coordinate compression:
    // Include each vertex coordinate and coordinate+1. Between two adjacent
    // compressed coordinates, the inside/outside status is uniform.
    // Build prefix sum of invalid compressed cells to quickly check rectangles.

    int *cxs = malloc(4 * n * sizeof(int));
    int *cys = malloc(4 * n * sizeof(int));
    int ncx = 0, ncy = 0;

    for (int i = 0; i < n; i++) {
        cxs[ncx++] = xs[i];
        cxs[ncx++] = xs[i] + 1;
        cys[ncy++] = ys[i];
        cys[ncy++] = ys[i] + 1;
    }
    qsort(cxs, ncx, sizeof(int), cmp_int);
    ncx = unique_arr(cxs, ncx);
    qsort(cys, ncy, sizeof(int), cmp_int);
    ncy = unique_arr(cys, ncy);

    // Build validity grid in compressed coordinates.
    // Each compressed cell (ci, cj) represents real tile at (cxs[ci], cys[cj]).
    int *valid = calloc(ncx * ncy, sizeof(int));

    for (int cj = 0; cj < ncy; cj++) {
        int py = cys[cj];
        for (int ci = 0; ci < ncx; ci++) {
            int px = cxs[ci];
            valid[cj * ncx + ci] = is_valid(px, py, xs, ys, n);
        }
    }

    // Find compressed indices for each red tile
    int *ci_red = malloc(n * sizeof(int));
    int *cj_red = malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        ci_red[i] = find_idx(cxs, ncx, xs[i]);
        cj_red[i] = find_idx(cys, ncy, ys[i]);
    }

    // 2D prefix sum of invalid cells (1 = invalid, 0 = valid)
    // prefix[(j+1)*(ncx+1)+(i+1)] = sum over [0..i] x [0..j]
    ll *prefix = calloc((ll)(ncx + 1) * (ncy + 1), sizeof(ll));

    for (int cj = 0; cj < ncy; cj++) {
        for (int ci = 0; ci < ncx; ci++) {
            ll inv = 1 - valid[cj * ncx + ci];
            prefix[(ll)(cj+1)*(ncx+1) + (ci+1)] = inv
                + prefix[(ll)cj*(ncx+1) + (ci+1)]
                + prefix[(ll)(cj+1)*(ncx+1) + ci]
                - prefix[(ll)cj*(ncx+1) + ci];
        }
    }

    // Query: count invalid compressed cells in [ci1..ci2] x [cj1..cj2] (inclusive)
    #define QUERY(ci1, ci2, cj1, cj2) \
        (prefix[(ll)(cj2+1)*(ncx+1) + (ci2+1)] \
        - prefix[(ll)(cj1)*(ncx+1) + (ci2+1)] \
        - prefix[(ll)(cj2+1)*(ncx+1) + (ci1)] \
        + prefix[(ll)(cj1)*(ncx+1) + (ci1)])

    // Check all pairs of red tiles; if their compressed rectangle is all-valid,
    // compute the real area.
    ll best2 = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (xs[i] == xs[j] && ys[i] == ys[j]) continue;

            int ci1 = ci_red[i], ci2 = ci_red[j];
            if (ci1 > ci2) { int t = ci1; ci1 = ci2; ci2 = t; }
            int cj1 = cj_red[i], cj2 = cj_red[j];
            if (cj1 > cj2) { int t = cj1; cj1 = cj2; cj2 = t; }

            if (QUERY(ci1, ci2, cj1, cj2) == 0) {
                ll dx = xs[j] - xs[i]; if (dx < 0) dx = -dx;
                ll dy = ys[j] - ys[i]; if (dy < 0) dy = -dy;
                ll area = (dx + 1) * (dy + 1);
                if (area > best2) best2 = area;
            }
        }
    }

    free(xs); free(ys);
    free(cxs); free(cys);
    free(valid);
    free(prefix);
    free(ci_red); free(cj_red);

    if (part == 2) {
        printf("Part 2: %lld\n", best2);
        return 0;
    }

    printf("Part 1: %lld\n", best1);
    printf("Part 2: %lld\n", best2);

    return 0;
}

File: benchmark/languages/go/work/day09/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	part := 0
	if len(os.Args) > 1 {
		part, _ = strconv.Atoi(os.Args[1])
	}

	f, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer f.Close()

	var reds [][2]int
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		reds = append(reds, [2]int{x, y})
	}

	if part == 0 || part == 1 {
		fmt.Println("Part 1:", part1(reds))
	}
	if part == 0 || part == 2 {
		fmt.Println("Part 2:", part2(reds))
	}
}

func part1(reds [][2]int) int {
	best := 0
	n := len(reds)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			dx := reds[i][0] - reds[j][0]
			if dx < 0 {
				dx = -dx
			}
			dy := reds[i][1] - reds[j][1]
			if dy < 0 {
				dy = -dy
			}
			area := (dx + 1) * (dy + 1)
			if area > best {
				best = area
			}
		}
	}
	return best
}

// For Part 2, we need to:
// 1. Compute all red+green tiles (polygon outline + interior)
// 2. Find the largest rectangle with red corners where all tiles are red/green
//
// Strategy:
// - Coordinate compress x and y values
// - In compressed space, mark cells that are entirely red/green
// - Use 2D prefix sums to check rectangles
// - For each pair of red tiles, check if the rectangle they define is fully covered

func part2(reds [][2]int) int {
	n := len(reds)

	// Collect all unique x and y coordinates from polygon edges
	// Each edge is horizontal or vertical between consecutive red tiles
	// We need to generate all relevant x/y coordinates for compression

	// Gather all x and y coordinates
	xSet := map[int]bool{}
	ySet := map[int]bool{}
	for _, r := range reds {
		xSet[r[0]] = true
		ySet[r[1]] = true
	}

	// For edges between consecutive red tiles:
	// The edges are axis-aligned, so we just need the endpoint coordinates
	// (already included above)

	// Build sorted unique coordinate arrays
	xs := sortedKeys(xSet)
	ys := sortedKeys(ySet)

	// Coordinate compression maps
	xIdx := map[int]int{}
	yIdx := map[int]int{}
	for i, v := range xs {
		xIdx[v] = i
	}
	for i, v := range ys {
		yIdx[v] = i
	}

	nx := len(xs)
	ny := len(ys)

	// Build a grid of "cells" between compressed coordinates.
	// We use a cell-based approach: cell (i,j) represents the region
	// [xs[i], xs[i+1]-1] x [ys[j], ys[j+1]-1] in real coordinates.
	// But for simplicity, let's think of the actual tile grid.
	//
	// Actually, let's think differently. Since all edges are axis-aligned and
	// endpoints are on our compressed coordinates, we can use a compressed grid
	// where each "cell" represents the block of tiles between consecutive
	// compressed coordinates.
	//
	// A compressed cell (i,j) for 0 <= i < nx-1, 0 <= j < ny-1 represents
	// real tiles x in [xs[i], xs[i+1]-1], y in [ys[j], ys[j+1]-1].
	// The "point" cell (i,j) where both are actual coordinates represents
	// exactly tile (xs[i], ys[j]).
	//
	// This gets complex. Let me use a simpler approach:
	// Since the polygon edges are axis-aligned, I'll use a different compressed
	// grid that captures both the edge lines and the gaps between them.
	//
	// Alternative: expand coordinates by 2 (2*x, 2*y) so we can represent
	// both tiles and half-gaps. But let's try a cleaner approach.

	// Simpler approach: use the "odd-even" rule.
	// For each compressed cell, determine if it's inside/on the polygon.
	// A compressed cell (ci, cj) corresponds to real coordinate (xs[ci], ys[cj]).
	// Wait - each compressed index is just a specific coordinate.
	// The "area" between consecutive coordinates we handle separately.

	// Let me think again. The compressed grid has nx * ny "points" where
	// each point is a real tile coordinate. Between consecutive compressed
	// x values, there may be many real tiles. Let's call those "gap cells".

	// The key insight: since all polygon edges are axis-aligned and their
	// endpoints are in our coordinate set, a horizontal strip between y[j] and y[j+1]
	// (for consecutive compressed y values) is entirely inside or outside the
	// polygon (except for the edge lines themselves).

	// I'll use a compressed grid with "segments" between coordinates:
	// - Pixel at compressed coords (ci, cj) = real tile (xs[ci], ys[cj])
	// - Horizontal segment between (ci, cj) and (ci+1, cj) = real tiles from xs[ci]+1 to xs[ci+1]-1 at y=ys[cj]
	// - Vertical segment between (ci, cj) and (ci, cj+1) = real tiles from ys[cj]+1 to ys[cj+1]-1 at x=xs[ci]
	// - Block between (ci,cj),(ci+1,cj),(ci,cj+1),(ci+1,cj+1) = interior block

	// This gives a 2*(nx) - 1 by 2*(ny) - 1 expanded compressed grid.

	// Let me use this expanded grid approach.
	// Expanded grid size: (2*nx-1) x (2*ny-1)
	// Mapping: real coordinate index (ci, cj) -> expanded index (2*ci, 2*cj)
	// Segment (ci, ci+1) at row cj -> expanded (2*ci+1, 2*cj)
	// Block -> expanded (2*ci+1, 2*cj+1)

	egx := 2*nx - 1
	egy := 2*ny - 1

	// Build edge set: mark which expanded cells are ON the polygon boundary
	edge := make([][]bool, egx)
	for i := range edge {
		edge[i] = make([]bool, egy)
	}

	// Mark red tiles
	for _, r := range reds {
		ci := xIdx[r[0]]
		cj := yIdx[r[1]]
		edge[2*ci][2*cj] = true
	}

	// Mark green tiles on edges (between consecutive red tiles)
	for i := 0; i < n; i++ {
		a := reds[i]
		b := reds[(i+1)%n]
		if a[0] == b[0] {
			// vertical edge
			ci := xIdx[a[0]]
			j1 := yIdx[a[1]]
			j2 := yIdx[b[1]]
			if j1 > j2 {
				j1, j2 = j2, j1
			}
			// Mark all tiles from ys[j1] to ys[j2] at x=xs[ci]
			// In expanded grid: x=2*ci, y from 2*j1 to 2*j2
			for ej := 2*j1 + 1; ej < 2*j2; ej++ {
				edge[2*ci][ej] = true
			}
		} else if a[1] == b[1] {
			// horizontal edge
			cj := yIdx[a[1]]
			i1 := xIdx[a[0]]
			i2 := xIdx[b[0]]
			if i1 > i2 {
				i1, i2 = i2, i1
			}
			// Mark all tiles from xs[i1] to xs[i2] at y=ys[cj]
			// In expanded grid: y=2*cj, x from 2*i1 to 2*i2
			for ei := 2*i1 + 1; ei < 2*i2; ei++ {
				edge[ei][2*cj] = true
			}
		}
	}

	// Now flood fill from outside to find all cells that are NOT inside the polygon
	// Use BFS from corners of the expanded grid
	inside := make([][]bool, egx)
	for i := range inside {
		inside[i] = make([]bool, egy)
	}

	// outside[i][j] = true means this expanded cell is outside (or on boundary but reachable from outside)
	outside := make([][]bool, egx)
	for i := range outside {
		outside[i] = make([]bool, egy)
	}

	type point struct{ x, y int }
	queue := []point{}

	enqueue := func(x, y int) {
		if x < 0 || x >= egx || y < 0 || y >= egy {
			return
		}
		if outside[x][y] || edge[x][y] {
			return
		}
		outside[x][y] = true
		queue = append(queue, point{x, y})
	}

	// Start flood fill from all border cells
	for i := 0; i < egx; i++ {
		enqueue(i, 0)
		enqueue(i, egy-1)
	}
	for j := 0; j < egy; j++ {
		enqueue(0, j)
		enqueue(egx-1, j)
	}

	dx := []int{0, 0, 1, -1}
	dy := []int{1, -1, 0, 0}

	for len(queue) > 0 {
		p := queue[0]
		queue = queue[1:]
		for d := 0; d < 4; d++ {
			nx2 := p.x + dx[d]
			ny2 := p.y + dy[d]
			enqueue(nx2, ny2)
		}
	}

	// A cell in expanded grid is "red or green" if it's an edge cell OR it's inside (not outside, not unreachable non-edge)
	// inside = !outside && !edge -> interior (green)
	// edge -> red or green (on boundary)
	// So red-or-green = edge OR (!outside && !edge) = !outside
	// Wait: edge cells are NOT marked as outside (we skip them in enqueue).
	// So a cell is red-or-green if: edge[x][y] OR (!outside[x][y] && !edge[x][y])
	// Which simplifies to: !outside[x][y] (since edge cells are not outside)
	// But wait: edge cells might not be outside, and non-edge non-outside cells are inside.
	// So: red-or-green = !outside[x][y]

	for i := 0; i < egx; i++ {
		for j := 0; j < egy; j++ {
			inside[i][j] = !outside[i][j]
		}
	}

	// Now I need to check: for a rectangle from (x1,y1) to (x2,y2) in real coordinates,
	// where (x1,y1) and (x2,y2) are both red tiles, is every tile in the rectangle inside?
	//
	// A rectangle in real space spans compressed indices:
	// x: from xIdx[x1] to xIdx[x2] (or vice versa)
	// y: from yIdx[y1] to yIdx[y2] (or vice versa)
	// In expanded space: x from 2*ci1 to 2*ci2, y from 2*cj1 to 2*cj2
	// (inclusive)
	//
	// For the rectangle to be all red/green, all expanded cells in that range must be inside.
	// We can use a 2D prefix sum on the "bad" cells (outside cells).

	// Build prefix sum of "bad" cells (cells that are outside)
	bad := make([][]int, egx+1)
	for i := range bad {
		bad[i] = make([]int, egy+1)
	}
	for i := 0; i < egx; i++ {
		for j := 0; j < egy; j++ {
			v := 0
			if outside[i][j] {
				v = 1
			}
			bad[i+1][j+1] = v + bad[i][j+1] + bad[i+1][j] - bad[i][j]
		}
	}

	queryBad := func(x1, y1, x2, y2 int) int {
		// Query sum of bad cells in [x1..x2] x [y1..y2] (0-indexed, inclusive)
		if x1 > x2 || y1 > y2 {
			return 0
		}
		return bad[x2+1][y2+1] - bad[x1][y2+1] - bad[x2+1][y1] + bad[x1][y1]
	}

	// For each pair of red tiles, compute the rectangle and check if valid
	best := 0
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			r1 := reds[i]
			r2 := reds[j]

			// Real coordinates of rectangle corners
			minX, maxX := r1[0], r2[0]
			if minX > maxX {
				minX, maxX = maxX, minX
			}
			minY, maxY := r1[1], r2[1]
			if minY > maxY {
				minY, maxY = maxY, minY
			}

			if minX == maxX || minY == maxY {
				// Degenerate rectangle (line) - area is (maxX-minX+1)*(maxY-minY+1)
				// Still valid as long as all tiles are red/green
				// Check using expanded grid
			}

			// Expanded grid indices
			ci1 := xIdx[minX]
			ci2 := xIdx[maxX]
			cj1 := yIdx[minY]
			cj2 := yIdx[maxY]

			ex1 := 2 * ci1
			ex2 := 2 * ci2
			ey1 := 2 * cj1
			ey2 := 2 * cj2

			// Check if all cells in expanded rectangle are "inside" (not outside)
			if queryBad(ex1, ey1, ex2, ey2) == 0 {
				// Valid rectangle
				area := (maxX - minX + 1) * (maxY - minY + 1)
				if area > best {
					best = area
				}
			}
		}
	}

	return best
}

func sortedKeys(m map[int]bool) []int {
	keys := make([]int, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Ints(keys)
	return keys
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

var _ = abs // suppress unused warning
